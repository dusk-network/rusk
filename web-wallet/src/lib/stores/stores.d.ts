type Readable<T> = import("svelte/store").Readable<T>;

type Writable<T> = import("svelte/store").Writable<T>;

type GasStoreContent = {
  gasLimitLower: bigint;
  gasLimitUpper: bigint;
  gasPriceLower: bigint;
};

type GasStore = Readable<GasStoreContent>;

type SettingsStoreContent = {
  currency: string;
  darkMode: boolean;
  dashboardTransactionLimit: number;
  gasLimit: bigint;
  gasPrice: bigint;
  hideStakingNotice: boolean;
  language: string;
  userId: string;
  walletCreationBlockHeight: bigint;
};

type SettingsStore = Writable<SettingsStoreContent> & {
  reset: () => void;
  resetGasSettings: () => void;
};

type TransactionInfo =
  | {
      hash: string;
      nullifiers: Uint8Array[];
    }
  | {
      hash: string;
      nonce: bigint;
    };

type TransactionsStoreContent = { transactions: Transaction[] };

type TransactionsStore = Readable<TransactionsStoreContent>;

type OperationsStoreContent = { currentOperation: string };

type OperationsStore = Writable<OperationsStoreContent>;

type NetworkStoreContent = {
  get connected(): boolean;
  networkName: string;
};

type NetworkSyncerOptions = {
  signal?: AbortSignal;
};

type NetworkStoreServices = {
  checkBlock: (height: bigint, hash: string) => Promise<boolean>;
  connect: () => Promise<import("$lib/vendor/w3sper.js/src/mod").Network>;
  disconnect: () => Promise<void>;
  getAccountSyncer: (
    options?: NetworkSyncerOptions
  ) => Promise<import("$lib/vendor/w3sper.js/src/mod").AccountSyncer>;
  getAddressSyncer: (
    options?: NetworkSyncerOptions
  ) => Promise<import("$lib/vendor/w3sper.js/src/mod").AddressSyncer>;
  getBlockHashByHeight: (height: bigint) => Promise<string>;
  getCurrentBlockHeight: () => Promise<bigint>;
  getLastFinalizedBlockHeight: () => Promise<bigint>;
  init: () => Promise<void>;
};

type NetworkStore = Readable<NetworkStoreContent> & NetworkStoreServices;

type NodeInfo = {
  bootstrappingNodes: Array<string>;
  chainId: number;
  kadcastAddress: string;
  version: string;
  versionBuild: string;
};

type WalletStoreBalance = {
  shielded: AddressBalance;
  unshielded: AccountBalance;
};

type WalletStoreContent = {
  balance: WalletStoreBalance;
  currentProfile: import("$lib/vendor/w3sper.js/src/mod").Profile | null;
  initialized: boolean;
  minimumStake: bigint;
  profiles: Array<import("$lib/vendor/w3sper.js/src/mod").Profile>;
  stakeInfo: StakeInfo;
  syncStatus: {
    from: bigint;
    isInProgress: boolean;
    last: bigint;
    error: Error | null;
    progress: number;
  };
};

type WalletStoreServices = {
  abortSync: () => void;

  clearLocalData: () => Promise<void>;

  clearLocalDataAndInit: (
    profileGenerator: import("$lib/vendor/w3sper.js/src/mod").ProfileGenerator,
    syncFromBlock?: bigint
  ) => Promise<void>;

  getTransactionsHistory: () => Promise<any>;

  init: (
    profileGenerator: import("$lib/vendor/w3sper.js/src/mod").ProfileGenerator,
    syncFromBlock?: bigint
  ) => Promise<void>;

  reset: () => void;

  setCurrentProfile: (
    profile: import("$lib/vendor/w3sper.js/src/mod").Profile
  ) => Promise<void>;

  shield: (
    amount: bigint,
    gas: import("$lib/vendor/w3sper.js/src/mod").Gas
  ) => Promise<TransactionInfo>;

  stake: (
    amount: bigint,
    gas: import("$lib/vendor/w3sper.js/src/mod").Gas
  ) => Promise<TransactionInfo>;

  sync: (fromBlock?: bigint) => Promise<void>;

  transfer: (
    to: string,
    amount: bigint,
    memo: string,
    gas: import("$lib/vendor/w3sper.js/src/mod").Gas
  ) => Promise<TransactionInfo>;

  unshield: (
    amount: bigint,
    gas: import("$lib/vendor/w3sper.js/src/mod").Gas
  ) => Promise<TransactionInfo>;

  unstake: (
    amount: bigint,
    gas: import("$lib/vendor/w3sper.js/src/mod").Gas
  ) => Promise<TransactionInfo>;

  claimRewards: (
    amount: bigint,
    gas: import("$lib/vendor/w3sper.js/src/mod").Gas
  ) => Promise<TransactionInfo>;
};

type WalletStore = Readable<WalletStoreContent> & WalletStoreServices;
