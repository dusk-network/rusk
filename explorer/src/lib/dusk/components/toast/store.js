import { get, writable } from "svelte/store";

/**
 * @type {import('svelte/store').Writable<ToastItem[]>}
 * @description Stores each toast as an object in the array
 */
export const toastList = writable([]);

/**
 * @type {import('svelte/store').Writable<Number>}
 * @description Stores the timer fused by all toasts
 */
export const toastTimer = writable(0);

/**
 * @param {string} message
 * @param {string} icon
 * @param {TooltipType} type
 */

function addToast(type, message, icon) {
  const id = `dusk-toast-${Math.random().toString(36)}`;

  toastList.update((store) => {
    return [
      ...store,
      {
        icon: icon,
        id: id,
        message: message,
        type: type,
      },
    ];
  });

  const timeoutID = window.setTimeout(() => {
    deleteToast(id, timeoutID);
  }, get(toastTimer));
}

/**
 * @param {string} id
 * @param {number} timeout
 */
function deleteToast(id, timeout) {
  window.clearTimeout(timeout);

  // Deletes toast from store queue
  toastList.update((store) => {
    return store.filter((toast) => toast.id !== id);
  });
}

export const toast = addToast;
