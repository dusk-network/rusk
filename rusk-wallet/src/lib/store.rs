// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// Copyright (c) DUSK NETWORK. All rights reserved.

use std::io::Read;
use std::{fmt, fs};
use std::path::PathBuf;
use bytes::{BytesMut, BufMut, Bytes, Buf};

use dusk_pki::{SecretSpendKey, SecretKey};
use dusk_wallet_core::{Store, derive_sk, derive_ssk};

use crate::lib::errors::CliError;

/// Default directory name to store settings and keystore
const DATA_DIR: &str = ".dusk";

/// Stores all the user's settings and keystore in the file system
pub struct LocalStore {
    name: String,
    path: PathBuf,
    seed: [u8; 64],
}

impl Store for LocalStore {

    type Error = CliError;

    /// Retrieves the seed used to derive keys.
    fn get_seed(&self) -> Result<[u8; 64], Self::Error> {
        Ok(self.seed)
    }

    /// Retrieves a derived secret spend key from the store.
    fn retrieve_ssk(&self, index: u64) -> Result<SecretSpendKey, Self::Error> {
        Ok(derive_ssk(&self.seed, index))
    }

    /// Retrieves a derived secret key from the store.
    fn retrieve_sk(&self, index: u64) -> Result<SecretKey, Self::Error> {
        Ok(derive_sk(&self.seed, index))
    }

}

impl LocalStore {

    /// Creates a new store
    pub fn new(name: String, seed: [u8;64]) -> Result<LocalStore, CliError> {

        // construct path to file
        let home = dirs::home_dir().unwrap();
        let mut path = PathBuf::new();
        path.push(home.as_os_str());
        path.push(DATA_DIR);
        path.push(&name);
        path = path.with_extension("dat");

        // get the seed
        let mut seed_bytes= [0u8; 64];
        seed_bytes.copy_from_slice(&seed);

        // create the local store
        let store = LocalStore{
            name: name,
            path: path,
            seed: seed_bytes,
        };

        // save to disk and return
        store.save()?;
        Ok(store)

    }


    /// Loads wallet file from file
    pub fn from_file(path: PathBuf, _pwd: String) -> Result<LocalStore, CliError> {

        // basic sanity check
        if !path.is_file() {
            return Err(CliError::CorruptedFile);
        }

        // attempt to load and decode wallet
        let data = Bytes::from(fs::read(&path).unwrap());
        let mut reader = data.reader();

        // wallet_seed
        let mut wallet_seed = [0u8; 64];
        reader.read(&mut wallet_seed)?;

        // extract the name
        let name = path.file_name()
            .ok_or(CliError::CorruptedFile)?.to_str()
            .ok_or(CliError::CorruptedFile)?;

        // create and return
        Ok(LocalStore{
            name: String::from(name),
            path: path,
            seed: wallet_seed,
        })

    }

    /// Saves wallet to a file
    pub fn save(&self) -> Result<(), CliError> {

        let mut buf = BytesMut::new();

        // wallet seed (todo: encrypt)
        buf.put_slice(&self.seed);

        // write file
        fs::write(&self.path, &buf[..]).unwrap();
        Ok(())

    }

    /// The name of the wallet
    pub fn name(&self) -> String {
        self.name.clone()
    }

}

impl fmt::Debug for LocalStore {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "-------------------\n\
            LocalStore: {}\n\
            Name: {}\n\
            Seed: {:?}",
            self.path.as_os_str().to_str().unwrap(), 
            self.name, self.seed)
    }
}