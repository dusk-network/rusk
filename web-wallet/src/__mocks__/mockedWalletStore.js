import { ProfileGenerator } from "$lib/vendor/w3sper.js/src/profile";

import { stakeInfo } from "$lib/mock-data";

import { mockReadableStore } from "$lib/dusk/test-helpers";

const seed = new Uint8Array(64);
const seeder = () => seed;
const profileGenerator = new ProfileGenerator(seeder);
const profiles = [
  await profileGenerator.default,
  await profileGenerator.next(),
  await profileGenerator.next(),
];
const currentProfile = profiles[0];
const shieldedBalance = {
  spendable: 50_000_000_000_000n,
  value: 2_345_000_000_000n,
};
const publicBalance = { nonce: 1234n, value: shieldedBalance.value / 2n };

/** @type {WalletStoreContent} */
const content = {
  balance: { publicBalance, shieldedBalance },
  currentProfile,
  initialized: true,
  minimumStake: 1_000_000_000_000n,
  profiles,
  stakeInfo,
  syncStatus: {
    error: null,
    from: 0n,
    isInProgress: false,
    last: 0n,
    progress: 0,
  },
};

const mockedWalletStore = mockReadableStore(content);

export default mockedWalletStore;
