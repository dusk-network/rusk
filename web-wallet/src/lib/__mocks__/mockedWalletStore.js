import { ProfileGenerator } from "$lib/vendor/w3sper.js/src/profile";

import { stakeInfo } from "$lib/mock-data";

import { mockReadableStore } from "$lib/dusk/test-helpers";

const seed = new Uint8Array(64);
const seeder = () => seed;
const profileGenerator = new ProfileGenerator(seeder);
const profiles = [
  await profileGenerator.default,
  await profileGenerator.next(),
  await profileGenerator.next(),
];

const currentProfile = profiles[0];
const shielded = { spendable: 50_000_000_000_000n, value: 2_345_000_000_000n };
const unshielded = { nonce: 1234n, value: shielded.value / 2n };

/** @type {WalletStoreContent & { currentProfile: NonNullable<WalletStoreContent["currentProfile"]> }} */
const content = {
  balance: { shielded, unshielded },
  currentProfile,
  initialized: true,
  minimumStake: 1_000_000_000_000n,
  profiles,
  stakeInfo,
  syncStatus: {
    error: null,
    from: 0n,
    isInProgress: false,
    last: 0n,
    progress: 0,
  },
};

const mockedWalletStore = mockReadableStore(content);

export default mockedWalletStore;
