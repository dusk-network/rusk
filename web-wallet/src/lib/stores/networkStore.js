import { writable } from "svelte/store";
import { browser } from "$app/environment";
import { always, condition, getKey, getPath, isUndefined, when } from "lamb";
import {
  AccountSyncer,
  AddressSyncer,
  Network,
  useAsProtocolDriver,
} from "@dusk/w3sper";

import { rejectAfter } from "$lib/dusk/promise";
import { makeNodeUrl } from "$lib/url";

import wasmPath from "$lib/vendor/dusk_wallet_core.wasm?url";

function getNetworkUrl() {
  if (browser) {
    return makeNodeUrl();
  } else {
    return (
      (import.meta.env.VITE_NODE_URL &&
        new URL(import.meta.env.VITE_NODE_URL)) ||
      new URL("https://localhost")
    );
  }
}

const networkUrl = getNetworkUrl();

/** @type {Network} */
const network = new Network(networkUrl);

/** @type {NetworkStoreContent} */
const initialState = {
  get connected() {
    return network.connected;
  },
  networkName: "unknown",
};

const networkStore = writable(initialState);
const { set, subscribe } = networkStore;

/**
 * Checks if a block with the given height and hash
 * exists on the network.
 *
 * @type {NetworkStoreServices["checkBlock"]}
 */
const checkBlock = (height, hash) =>
  connect()
    .then(() => network.query(`checkBlock(height: ${height}, hash: "${hash}")`))
    .then(getKey("checkBlock"));

/** @type {NetworkStoreServices["connect"]} */
const connect = async () =>
  network.connected
    ? network
    : Promise.race([
        fetch(wasmPath)
          .then((response) => response.arrayBuffer())
          .then((buffer) => {
            useAsProtocolDriver(new Uint8Array(buffer));

            return network.connect();
          }),
        rejectAfter(
          10000,
          new Error("Timed out while connecting to the network")
        ),
      ]);

/** @type {NetworkStoreServices["disconnect"]} */
const disconnect = () => network.disconnect();

/** @type {() => Promise<AccountSyncer>} */
const getAccountSyncer = () => connect().then(() => new AccountSyncer(network));

/** @type {() => Promise<AddressSyncer>} */
const getAddressSyncer = () => connect().then(() => new AddressSyncer(network));

/** @type {NetworkStoreServices["getBlockHashByHeight"]} */
const getBlockHashByHeight = (height) =>
  connect()
    .then(() => network.query(`block(height: ${height}) { header { hash } }`))
    .then(getPath("block.header.hash"))
    .then(when(isUndefined, always("")));

/** @type {NetworkStoreServices["getCurrentBlockHeight"]} */
const getCurrentBlockHeight = () => network.blockHeight;

/** @type {NetworkStoreServices["getLastFinalizedBlockHeight"]} */
const getLastFinalizedBlockHeight = () =>
  connect()
    .then(() => network.query("lastBlockPair { json }"))
    .then(getPath("lastBlockPair.json.last_finalized_block.0"))
    .then(condition(isUndefined, always(0n), BigInt));

/** @type {NetworkStoreServices["init"]} */
async function init() {
  const info = await network.node.info;

  set({
    ...initialState,
    networkName: info.chain.toString(),
  });
}

/** @type {NetworkStore} */
export default {
  checkBlock,
  connect,
  disconnect,
  getAccountSyncer,
  getAddressSyncer,
  getBlockHashByHeight,
  getCurrentBlockHeight,
  getLastFinalizedBlockHeight,
  init,
  subscribe,
};
