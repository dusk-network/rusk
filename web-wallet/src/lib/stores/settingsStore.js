import { writable } from "svelte/store";
import { browser } from "$app/environment";

// =========================
// --- Utilities ---
// =========================

export const serializeProperty = (/** @type {any} */ value) =>
  JSON.stringify(value, (_, v) => (typeof v === "bigint" ? `${v}n` : v));

export const deserializeProperty = (/** @type {string} */ value) =>
  JSON.parse(value, (_, v) =>
    typeof v === "string" && /^\d+n$/.test(v) ? BigInt(v.slice(0, -1)) : v
  );

// =========================
// --- End of Utilities ---
// =========================

const browserDefaults = browser
  ? {
      darkMode: window.matchMedia("(prefers-color-scheme: dark)").matches,
      language: navigator.language,
    }
  : {
      darkMode: false,
      language: "en",
    };

/** @type {SettingsStoreContent} */
const initialState = {
  ...browserDefaults,
  currency: "USD",
  dashboardTransactionLimit: 5,
  gasLimit: BigInt(import.meta.env.VITE_GAS_LIMIT_DEFAULT ?? 100_000_000),
  gasPrice: BigInt(import.meta.env.VITE_GAS_PRICE_DEFAULT ?? 1),
  hideStakingNotice: false,
  userId: "",
  walletCreationBlockHeight: 0n,
};

const createPersistedStore = (
  /** @type {string} */ key,
  /** @type {SettingsStoreContent} */ initialValue
) => {
  const load = () => {
    if (!browser) {
      return initialValue;
    }
    const storedValue = localStorage.getItem(key);
    return storedValue ? deserializeProperty(storedValue) : initialValue;
  };

  const store = writable(load());

  if (browser) {
    store.subscribe((value) => {
      localStorage.setItem(key, serializeProperty(value));
    });
  }

  return store;
};

const settingsStore = createPersistedStore(
  `${CONFIG.LOCAL_STORAGE_APP_KEY}-preferences`,
  initialState
);

const { set, subscribe, update } = settingsStore;

// Reset store to initial state
const reset = () => set(initialState);

// Resets only gas settings to their defaults.
const resetGasSettings = () =>
  update((current) => ({
    ...current,
    gasLimit: initialState.gasLimit,
    gasPrice: initialState.gasPrice,
  }));

/** @type {SettingsStore} */
export default {
  reset,
  resetGasSettings,
  set,
  subscribe,
  update,
};
