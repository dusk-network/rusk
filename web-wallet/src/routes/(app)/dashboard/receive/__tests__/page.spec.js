import { afterAll, afterEach, describe, expect, it, vi } from "vitest";
import { cleanup, fireEvent, render } from "@testing-library/svelte";
import { get } from "svelte/store";

import mockedWalletStore from "$lib/__mocks__/mockedWalletStore";

import { walletStore } from "$lib/stores";

import Receive from "../+page.svelte";

vi.useFakeTimers();

vi.mock("$lib/dusk/string", async (importOriginal) => {
  /** @type {typeof import("$lib/dusk/string")} */
  const original = await importOriginal();

  return {
    ...original,
    randomUUID: () => "some-generated-id",
  };
});

vi.mock("$lib/stores", async (importOriginal) => {
  /** @type {WalletStore} */
  const original = await importOriginal();

  return {
    ...original,
    walletStore: {
      ...mockedWalletStore,
    },
  };
});

describe("Receive", () => {
  const currentProfile =
    /** @type {import("$lib/vendor/w3sper.js/src/mod").Profile} */ (
      get(walletStore).currentProfile
    );
  const expectedAddress = currentProfile.address.toString();
  const expectedAccount = currentProfile.account.toString();

  afterEach(cleanup);

  afterAll(() => {
    vi.doUnmock("$lib/dusk/string");
    vi.doUnmock("$lib/stores");
    vi.useRealTimers();
  });

  it("should render the receive page with a choice to switch from public to shielded address", async () => {
    const { container, getByRole, getByText } = render(Receive);

    await vi.runAllTimersAsync();

    expect(getByRole("radiogroup")).toBeInTheDocument();
    expect(getByText(expectedAccount)).toBeInTheDocument();
    expect(() => getByText(expectedAddress)).toThrow();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should show the shielded address when the user makes such choice", async () => {
    const { container, getByRole, getByText } = render(Receive);

    await vi.runAllTimersAsync();

    expect(getByText(expectedAccount)).toBeInTheDocument();
    expect(() => getByText(expectedAddress)).toThrow();

    await fireEvent.click(getByRole("radio", { checked: false }));
    await vi.runAllTimersAsync();

    expect(() => getByText(expectedAccount)).toThrow();
    expect(getByText(expectedAddress)).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });
});
