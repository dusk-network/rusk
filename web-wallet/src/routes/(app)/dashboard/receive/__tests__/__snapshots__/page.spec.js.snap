// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Receive > should change the icon to "public" and show the public address when the user makes such choice 1`] = `
<div
  class="dusk-card dusk-card--gap-medium"
>
  <div
    class="dusk-card__header-container"
  >
    <header
      class="dusk-card__header"
      slot="header"
    >
      <div
        class="dusk-card__heading-container"
      >
        <h3
          class="h4"
        >
          Receive
        </h3>
         
        <div
          class="dusk-card__heading-icons"
        >
          
          <svg
            class="dusk-icon dusk-icon--size--default"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M12.07 19L9.24 16.18L19 6.42L17.58 5L7.82 14.76L5 11.94V19Z"
            />
          </svg>
          
          
          <svg
            class="dusk-icon dusk-icon--size--default"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 1L3 5V11C3 16.5 6.8 21.7 12 23C17.2 21.7 21 16.5 21 11V5L12 1M16 15.8C16 16.4 15.4 17 14.7 17H9.2C8.6 17 8 16.4 8 15.7V12.2C8 11.6 8.6 11 9.2 11V8.5C9.2 7.1 10.6 6 12 6S14.8 7.1 14.8 8.5V9H13.5V8.5C13.5 7.7 12.8 7.2 12 7.2S10.5 7.7 10.5 8.5V11H14.8C15.4 11 16 11.6 16 12.3V15.8Z"
            />
          </svg>
          
        </div>
      </div>
    </header>
  </div>
   
  <div
    class="dusk-card__body-container"
  >
    <div
      class="dusk-exclusive-choice"
      role="radiogroup"
    >
      <input
        class="dusk-exclusive-choice__radio"
        id="some-generated-id-shielded"
        name="some-generated-id"
        type="radio"
        value="shielded"
      />
       
      <label
        class="dusk-exclusive-choice__label"
        for="some-generated-id-shielded"
      >
        Shielded
      </label>
      <input
        class="dusk-exclusive-choice__radio"
        id="some-generated-id-public"
        name="some-generated-id"
        type="radio"
        value="public"
      />
       
      <label
        class="dusk-exclusive-choice__label"
        for="some-generated-id-public"
      >
        Public
      </label>
    </div>
     
    <div
      class="receive svelte-14g7par"
    >
      <figure
        class="receive__address-qr-figure svelte-14g7par"
      >
        <img
          alt="Key QR code"
          class="dusk-qr-code receive__qr"
          height="0"
          src="data:image/png;base64,00"
          width="0"
        />
        
         
        <figcaption
          class="receive__address svelte-14g7par"
        >
          <samp>
            bEjRbupbXK9SFT43T4H46nvD2hC7hhsQM1BGKqM5sHxzBzATF4swNEbjJympWpiWufQ5qw9xPErsxnQcHicp8XKoYiYbGv1DFTK5zJV4yQ1VzMM8H3uXvqbZKiBwmzGeJ8J
          </samp>
        </figcaption>
      </figure>
       
      <div
        class="receive__buttons svelte-14g7par"
      >
        <a
          aria-disabled="false"
          class="dusk-anchor dusk-anchor--on-surface dusk-anchor-button dusk-anchor-button--variant--tertiary dusk-anchor-button--size--default dusk-icon-button--labeled receive__button"
          href="/some-base-path/dashboard"
        >
          <svg
            class="dusk-icon dusk-icon--size--default dusk-anchor-button__icon"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M20,11V13H8L13.5,18.5L12.08,19.92L4.16,12L12.08,4.08L13.5,5.5L8,11H20Z"
            />
          </svg>
          
           
          <span
            class="dusk-anchor-button__text"
          >
            Back
          </span>
          
          
        </a>
         
        <button
          class="dusk-button dusk-button--type--button dusk-button--variant--primary dusk-button--size--default dusk-icon-button--labeled receive__button"
          type="button"
        >
          <svg
            class="dusk-icon dusk-icon--size--default dusk-button__icon"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"
            />
          </svg>
          
           
          <span
            class="dusk-button__text"
          >
            Copy
          </span>
          
        </button>
      </div>
    </div>
  </div>
   
</div>
`;

exports[`Receive > should render the receive page with a double icon and a choice to switch from shielded to public address 1`] = `
<div
  class="dusk-card dusk-card--gap-medium"
>
  <div
    class="dusk-card__header-container"
  >
    <header
      class="dusk-card__header"
      slot="header"
    >
      <div
        class="dusk-card__heading-container"
      >
        <h3
          class="h4"
        >
          Receive
        </h3>
         
        <div
          class="dusk-card__heading-icons"
        >
          
          <svg
            class="dusk-icon dusk-icon--size--default"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M12.07 19L9.24 16.18L19 6.42L17.58 5L7.82 14.76L5 11.94V19Z"
            />
          </svg>
          
          
          <svg
            class="dusk-icon dusk-icon--size--default"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M12,1L3,5V11C3,16.55 6.84,21.74 12,23C17.16,21.74 21,16.55 21,11V5L12,1M12,7C13.4,7 14.8,8.1 14.8,9.5V11C15.4,11 16,11.6 16,12.3V15.8C16,16.4 15.4,17 14.7,17H9.2C8.6,17 8,16.4 8,15.7V12.2C8,11.6 8.6,11 9.2,11V9.5C9.2,8.1 10.6,7 12,7M12,8.2C11.2,8.2 10.5,8.7 10.5,9.5V11H13.5V9.5C13.5,8.7 12.8,8.2 12,8.2Z"
            />
          </svg>
          
        </div>
      </div>
    </header>
  </div>
   
  <div
    class="dusk-card__body-container"
  >
    <div
      class="dusk-exclusive-choice"
      role="radiogroup"
    >
      <input
        class="dusk-exclusive-choice__radio"
        id="some-generated-id-shielded"
        name="some-generated-id"
        type="radio"
        value="shielded"
      />
       
      <label
        class="dusk-exclusive-choice__label"
        for="some-generated-id-shielded"
      >
        Shielded
      </label>
      <input
        class="dusk-exclusive-choice__radio"
        id="some-generated-id-public"
        name="some-generated-id"
        type="radio"
        value="public"
      />
       
      <label
        class="dusk-exclusive-choice__label"
        for="some-generated-id-public"
      >
        Public
      </label>
    </div>
     
    <div
      class="receive svelte-14g7par"
    >
      <figure
        class="receive__address-qr-figure svelte-14g7par"
      >
        <img
          alt="Key QR code"
          class="dusk-qr-code receive__qr"
          height="0"
          src="data:image/png;base64,00"
          width="0"
        />
        
         
        <figcaption
          class="receive__address svelte-14g7par"
        >
          <samp>
            2zFZwXf4JEvD2Vmd4kqx5tHDv73LkRr2Ydb84Knr2gmFJJoEEpZPtTbuQCrojM85QDHAvosxoZdnpLPSNRqzqmai
          </samp>
        </figcaption>
      </figure>
       
      <div
        class="receive__buttons svelte-14g7par"
      >
        <a
          aria-disabled="false"
          class="dusk-anchor dusk-anchor--on-surface dusk-anchor-button dusk-anchor-button--variant--tertiary dusk-anchor-button--size--default dusk-icon-button--labeled receive__button"
          href="/some-base-path/dashboard"
        >
          <svg
            class="dusk-icon dusk-icon--size--default dusk-anchor-button__icon"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M20,11V13H8L13.5,18.5L12.08,19.92L4.16,12L12.08,4.08L13.5,5.5L8,11H20Z"
            />
          </svg>
          
           
          <span
            class="dusk-anchor-button__text"
          >
            Back
          </span>
          
          
        </a>
         
        <button
          class="dusk-button dusk-button--type--button dusk-button--variant--primary dusk-button--size--default dusk-icon-button--labeled receive__button"
          type="button"
        >
          <svg
            class="dusk-icon dusk-icon--size--default dusk-button__icon"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"
            />
          </svg>
          
           
          <span
            class="dusk-button__text"
          >
            Copy
          </span>
          
        </button>
      </div>
    </div>
  </div>
   
</div>
`;

exports[`Receive > should render the receive page with a single icon and no choice for shielded / public if the feature flag is not true 1`] = `
<div
  class="dusk-card dusk-card--gap-medium"
>
  <div
    class="dusk-card__header-container"
  >
    <header
      class="dusk-card__header"
      slot="header"
    >
      <div
        class="dusk-card__heading-container"
      >
        <h3
          class="h4"
        >
          Receive
        </h3>
         
        <div
          class="dusk-card__heading-icons"
        >
          
          <svg
            class="dusk-icon dusk-icon--size--default"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M12.07 19L9.24 16.18L19 6.42L17.58 5L7.82 14.76L5 11.94V19Z"
            />
          </svg>
          
        </div>
      </div>
    </header>
  </div>
   
  <div
    class="dusk-card__body-container"
  >
     
    <div
      class="receive svelte-14g7par"
    >
      <figure
        class="receive__address-qr-figure svelte-14g7par"
      >
        <img
          alt="Key QR code"
          class="dusk-qr-code receive__qr"
          height="0"
          src="data:image/png;base64,00"
          width="0"
        />
        
         
        <figcaption
          class="receive__address svelte-14g7par"
        >
          <samp>
            2zFZwXf4JEvD2Vmd4kqx5tHDv73LkRr2Ydb84Knr2gmFJJoEEpZPtTbuQCrojM85QDHAvosxoZdnpLPSNRqzqmai
          </samp>
        </figcaption>
      </figure>
       
      <div
        class="receive__buttons svelte-14g7par"
      >
        <a
          aria-disabled="false"
          class="dusk-anchor dusk-anchor--on-surface dusk-anchor-button dusk-anchor-button--variant--tertiary dusk-anchor-button--size--default dusk-icon-button--labeled receive__button"
          href="/some-base-path/dashboard"
        >
          <svg
            class="dusk-icon dusk-icon--size--default dusk-anchor-button__icon"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M20,11V13H8L13.5,18.5L12.08,19.92L4.16,12L12.08,4.08L13.5,5.5L8,11H20Z"
            />
          </svg>
          
           
          <span
            class="dusk-anchor-button__text"
          >
            Back
          </span>
          
          
        </a>
         
        <button
          class="dusk-button dusk-button--type--button dusk-button--variant--primary dusk-button--size--default dusk-icon-button--labeled receive__button"
          type="button"
        >
          <svg
            class="dusk-icon dusk-icon--size--default dusk-button__icon"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"
            />
          </svg>
          
           
          <span
            class="dusk-button__text"
          >
            Copy
          </span>
          
        </button>
      </div>
    </div>
  </div>
   
</div>
`;
